### ISSUE TITLE
Task 3: Implement Transformer architecture for attention-based forecasting

### ISSUE THREAD

**Issue by jpequegn** [jpequegn/ts-rs#49]
# Task 3: Implement Transformer Architecture for Attention-Based Forecasting

**Related to Epic**: #46  
**Depends on**: #47

## Objective
Implement Transformer neural networks with self-attention mechanisms optimized for time series forecasting, enabling the model to capture long-range dependencies and complex temporal patterns.

## Scope
- Create `src/ml/transformer.rs` with Transformer implementation
- Implement self-attention and multi-head attention mechanisms
- Add positional encoding for time series data
- Support for both encoder-only and encoder-decoder architectures

### Core Functions to Implement
```rust
// Main transformer interface
pub fn create_transformer_forecaster(
    config: &TransformerConfig,
    training_data: &TimeSeries,
) -> Result<TransformerForecaster, MLError>;

// Attention mechanisms
pub fn multi_head_attention(
    query: &Tensor,
    key: &Tensor,
    value: &Tensor,
    num_heads: usize,
    mask: Option<&Tensor>,
) -> Result<Tensor, MLError>;

// Training and prediction
pub fn train_transformer_model(
    model: &mut TransformerForecaster,
    data: &TimeSeriesDataset,
    config: &TrainingConfig,
) -> Result<TrainingResult, MLError>;
```

### Data Structures
```rust
pub struct TransformerForecaster {
    pub encoder: TransformerEncoder,
    pub decoder: Option<TransformerDecoder>,
    pub positional_encoding: PositionalEncoding,
    pub input_projection: DenseLayer,
    pub output_projection: DenseLayer,
    pub config: TransformerConfig,
}

pub struct TransformerConfig {
    pub d_model: usize,              // Model dimension
    pub n_heads: usize,              // Number of attention heads
    pub d_ff: usize,                 // Feed-forward dimension
    pub n_encoder_layers: usize,     // Number of encoder layers
    pub n_decoder_layers: Option<usize>, // Number of decoder layers
    pub dropout_rate: f32,
    pub max_sequence_length: usize,
    pub attention_dropout: f32,
    pub activation: ActivationType,
    pub layer_norm_eps: f64,
}

pub struct TransformerEncoder {
    pub layers: Vec<TransformerEncoderLayer>,
    pub layer_norm: LayerNorm,
}

pub struct TransformerEncoderLayer {
    pub self_attention: MultiHeadAttention,
    pub feed_forward: FeedForwardNetwork,
    pub layer_norm1: LayerNorm,
    pub layer_norm2: LayerNorm,
    pub dropout: Dropout,
}

pub struct MultiHeadAttention {
    pub num_heads: usize,
    pub d_model: usize,
    pub d_k: usize,  // Key/Query dimension
    pub d_v: usize,  // Value dimension
    pub w_q: DenseLayer,  // Query projection
    pub w_k: DenseLayer,  // Key projection
    pub w_v: DenseLayer,  // Value projection
    pub w_o: DenseLayer,  // Output projection
    pub attention_dropout: Dropout,
}
```

### 1. Self-Attention Mechanism
- **Scaled Dot-Product Attention**: Core attention computation
- **Multi-Head Attention**: Parallel attention in different representation subspaces
- **Causal Masking**: Prevent future information leakage during training
- **Attention Weights**: Extract and visualize attention patterns

### 2. Positional Encoding
```rust
pub struct PositionalEncoding {
    pub encoding_type: PositionalEncodingType,
    pub max_length: usize,
    pub d_model: usize,
}

pub enum PositionalEncodingType {
    Sinusoidal,           // Standard sinusoidal encoding
    Learned,              // Learnable positional embeddings
    Relative,             // Relative position encoding
    Temporal,             // Time-aware encoding for irregular intervals
}

impl PositionalEncoding {
    pub fn encode_timestamps(&self, timestamps: &[DateTime<Utc>]) -> Tensor {
        // Convert timestamps to positional encoding
    }
    
    pub fn encode_sequence(&self, sequence_length: usize) -> Tensor {
        // Standard sequence position encoding
    }
}
```

### 3. Transformer Architecture Variants
- **Encoder-Only**: For time series classification and feature extraction
- **Encoder-Decoder**: For sequence-to-sequence forecasting
- **Decoder-Only**: For autoregressive generation
- **Sparse Attention**: Efficient attention for very long sequences

### 4. Time Series Specific Adaptations
```rust
// Time series specific transformer layer
pub struct TimeSeriesTransformerLayer {
    pub temporal_attention: TemporalAttention,
    pub seasonal_attention: SeasonalAttention,
    pub trend_projection: DenseLayer,
    pub feature_attention: FeatureAttention,
}

// Attention mechanisms specialized for time series
pub struct TemporalAttention {
    pub window_size: usize,
    pub stride: usize,
    pub attention_type: TemporalAttentionType,
}

pub enum TemporalAttentionType {
    Local,      // Local window attention
    Dilated,    // Dilated attention for different scales
    Hierarchical, // Multi-scale hierarchical attention
}
```

#### Attention Visualization and Interpretability
```rust
pub struct AttentionAnalysis {
    pub attention_weights: Vec<Tensor>,
    pub head_contributions: Vec<f32>,
    pub layer_contributions: Vec<f32>,
    pub temporal_focus: Vec<f32>,
}

impl TransformerForecaster {
    pub fn analyze_attention(&self, input: &Tensor) -> AttentionAnalysis {
        // Extract and analyze attention patterns
    }
    
    pub fn get_feature_importance(&self, input: &Tensor) -> Vec<f32> {
        // Compute feature importance via attention
    }
}
```

#### Memory Optimization
- **Gradient Checkpointing**: Trade computation for memory
- **Mixed Precision**: Use FP16 for memory efficiency
- **Sparse Attention**: Reduce memory complexity for long sequences
- **Layer Sharing**: Share parameters across layers

### 6. Integration with Forecasting Pipeline
```rust
impl ForecastMethod for TransformerForecaster {
    fn forecast(&self, data: &TimeSeries, horizon: usize) -> Result<ForecastResult, MLError> {
        // Transformer-specific forecasting with attention
        let input_sequence = self.prepare_input(data)?;
        let predictions = self.forward(input_sequence, horizon)?;
        let confidence = self.compute_prediction_confidence(&predictions)?;
        
        Ok(ForecastResult {
            predictions,
            confidence_intervals: Some(confidence),
            attention_weights: Some(self.extract_attention_weights()),
        })
    }
}
```

### 1. Efficient Attention Implementation
- **Flash Attention**: Memory-efficient attention computation
- **Chunked Attention**: Process attention in chunks for large sequences
- **Attention Caching**: Cache attention computations for inference
- **Kernel Fusion**: Fuse operations for GPU efficiency

### 2. Training Optimizations
- **Gradient Accumulation**: Handle large batch sizes with limited memory
- **Learning Rate Warmup**: Gradual learning rate increase for stable training
- **Layer-wise Learning Rates**: Different learning rates for different layers
- **Loss Scaling**: Prevent underflow in mixed precision training

## Acceptance Criteria
- [ ] Transformer models achieve competitive accuracy on time series benchmarks
- [ ] Self-attention mechanism correctly captures temporal dependencies
- [ ] Positional encoding works with both regular and irregular time series
- [ ] Memory usage is reasonable for sequences up to 2048 timesteps
- [ ] Attention weights are interpretable and meaningful
- [ ] GPU acceleration provides significant speedup over CPU
- [ ] Integration with existing forecasting framework is seamless
- [ ] Comprehensive test suite with attention mechanism validation

## Test Cases Required
- Long sequence forecasting (>1000 timesteps)
- Multi-variate time series with complex interactions
- Irregular time series with varying intervals
- Attention pattern visualization and interpretation
- Memory usage with very long sequences
- Comparison with LSTM/GRU on same datasets

## Dependencies
- Task 1 results for core ML infrastructure
- `candle-core` and `candle-nn` for neural network operations
- `candle-transformers` for transformer-specific layers
- Existing `forecasting` module for integration
- Linear algebra libraries for efficient matrix operations

## Estimated Time
20-25 hours

## Next Steps
After completion, enables:
- Task 4: Ensemble methods combining Transformers with other models
- Task 5: AutoML for automatic architecture search
- Task 6: Advanced interpretability and explanation methods
- Integration with existing anomaly detection for attention-based anomaly scoring
